File: Algorithms/24. Swap Nodes in Pairs.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn swap_pairs(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        match head {
            None => None,
            Some(mut inner_h) => match inner_h.next {
                None => Some(inner_h),
                Some(mut inner_s) => {
                    inner_h.next = Solution::swap_pairs(inner_s.next);
                    inner_s.next = Some(inner_h);
                    Some(inner_s)
                }
            }
        }
    }
}

File: Algorithms/24. Swap Nodes in Pairs.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function swapPairs(head: ListNode | null): ListNode | null {
  const dummy = new ListNode(-1, head);
  let cur = dummy;
  while (cur.next && cur.next.next) {
    const left = cur.next;
    const right = cur.next.next;
    [left.next, right.next, cur.next] = [right.next, left, right];
    cur = cur.next.next;
  }
  return dummy.next;
}
